lib/adm.js

```diff
@@ -4097,6 +4121,106 @@ maAdm.prototype.generatePlan = function (opts, callback)
 	setTimeout(callback, 0);
 };
 
+/*
+ * Determine whether the plan that was generated adheres to any restrictions
+ * that are in place.
+ * For now the only check implemented is to ensure that all images for
+ * deployment are available on the specific imgapi channel set by
+ * 'sdcadm channel set' on the 'remote_imgapi' set in etc/config.json.
+ *
+ * Options:
+ * - skip_verify_channel: if true, we avoid verifying that each image to be
+ *   provisioned is present on the defined SDC updates.joyent.com channel.
+ */
+maAdm.prototype.verifyPlan = function (opts, cb)
+{
+	assert.ok(this.ma_plan !== null);
+	assertplus.object(opts, 'opts');
+	assertplus.func(cb, 'cb');
+	assertplus.bool(opts.skip_verify_channel, 'opts.skip_verify_channel');
+
+	var self = this;
+
+	if (opts.skip_verify_channel) {
+		return (cb());
+	}
+	/*
+	 * If we're doing channel verification, go through our plan,
+	 * checking that each new image exists on that channel. To do this,
+	 * find the set of new image uuids per-service across all compute nodes
+	 * in the plan.
+	 */
+	// We don't have a Set() object, so instead use unique Object keys
+	var services_by_image = {};
+	for (var svcname in this.ma_plan) {
+		for (var cnid in this.ma_plan[svcname]) {
+			for (var i = 0;
+			    i < this.ma_plan[svcname][cnid].length;
+			    i++) {
+				var plan_op = this.ma_plan[svcname][cnid][i];
+				var new_image = null;
+				if (plan_op.action === 'provision') {
+					new_image = plan_op.IMAGE;
+				} else if (plan_op.action === 'reprovision') {
+					new_image = plan_op.new_image;
+				}
+				if (new_image) {
+					services_by_image[new_image] = svcname;
+				}
+			}
+		}
+	}
+
+	var imgapi_opts = {'channel': self.ma_channel};
+
+	// An array of strings, later used to form an error message.
+	var imagesNotInTheChannel = [];
+	// process each of the images in series
+	vasync.forEachParallel({
+	    inputs: Object.keys(services_by_image),
+	    func: function channelCheckOneImage(imageUuid, nextImage) {
+			self.ma_log.debug(
```

Comment:
Getting a little squished to the right here. :)  Is the body of channelCheckOneImage indented one more tab than wanted?

(Generally all whitespace commentary can be totally ignored by the author, including this one if you like!)

Reply:
Yep, I'll move that back out a level, to match the style on 4450, `func': function execPlanSvc(svcname, subcb) {`

I don't completely understand indentation rules here, but happy to go with the flow.